### PR TITLE
feat: added facenet512 onnx support for fast inference on gpu

### PR DESCRIPTION
**Pull request by ShivamSinghal1**
### What has been done

With this PR, facenet512 onnx support is added

## How to test

```shell
make lint && make test
```

### CODE REVIEW

File: tests/test_verify.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 logger = Logger()
 
-models = ["VGG-Face", "Facenet", "Facenet512", "ArcFace", "GhostFaceNet"]
+models = ["VGG-Face", "Facenet", "Facenet512", "ArcFace", "GhostFaceNet", "Facenet512ONNX"]
```
Comment (serengil) [inline tests/test_verify.py:11]: Seems you added an optional facial recognition model. If you add it here, it will get an exception in the pipeline. So, optional models should not be tested.

File: deepface/models/facial_recognition/Facenet.py [hunk 2]
```diff
@@ -1711,3 +1731,36 @@ def load_facenet512d_model(
     # -------------------------
 
     return model
+
+
+def load_facenet512d_onnx_model(
+    url="https://github.com/ShivamSinghal1/deepface/releases/download/v1/facenet512_fp32.onnx",
+) -> Model:
```
Comment (serengil) [inline deepface/models/facial_recognition/Facenet.py:1738]: Is this function really retuning keras.models.Model?

File: tests/test_verify.py (hunk 1 above)
Comment (serengil) [reply tests/test_verify.py:11]: As i mentioned, pipeline failed - https://github.com/serengil/deepface/actions/runs/10579001173/job/29310896976?pr=1322

**Comment by serengil** [comment]
Did you convert this weights to onnx? - https://github.com/serengil/deepface/blob/master/deepface/models/facial_recognition/Facenet.py#L1688

Should I wait same accuracy?

**Comment by serengil** [comment]
For the documentation, would you please describe the pros of onnx? We already have FaceNet512d model, why should I adopt this?

**Comment by ShivamSinghal1** [comment]
> Did you convert this weights to onnx? - https://github.com/serengil/deepface/blob/master/deepface/models/facial_recognition/Facenet.py#L1688
> 
> Should I wait same accuracy?

yes i have used this weights to convert it into onnx

File: deepface/models/facial_recognition/Facenet.py (hunk 2 above)
Comment (ShivamSinghal1) [reply deepface/models/facial_recognition/Facenet.py]: nope i have changed it to Any

File: tests/test_verify.py (hunk 1 above)
Comment (ShivamSinghal1) [reply tests/test_verify.py]: okay i have removed it, but tested on my end

File: deepface/models/facial_recognition/Facenet.py [hunk 3]
```diff
@@ -67,6 +69,26 @@ def __init__(self):
...
+        self.model_name = "FaceNet-512d-onnx"
+        self.input_shape = (160, 160)
+        self.output_shape = 512
+
+    def forward(self, img: np.ndarray) -> List[float]:
+        input_name = self.model.get_inputs()[0].name
+        output_name = self.model.get_outputs()[0].name
+        result = self.model.run([output_name], {input_name: img})
+        if isinstance(result[0][0], np.ndarray):
+            return result[0][0].tolist()
```
Comment (serengil) [inline deepface/models/facial_recognition/Facenet.py:88]: result[0][0] should always be numpy array, you can get rid of if block and return result[0][0].tolist() if i am not wrong

File: deepface/models/facial_recognition/Facenet.py [hunk 4]
```diff
@@ -1711,3 +1733,36 @@ def load_facenet512d_model(
...
     return model
+
+
+def load_facenet512d_onnx_model(
+    url="https://github.com/ShivamSinghal1/deepface/releases/download/v1/facenet512_fp32.onnx",
+) -> Any:
+    """
+    Download Facenet512d ONNX model weights and load
+    Returns:
+        model (Model)
```
Comment (serengil) [inline deepface/models/facial_recognition/Facenet.py:1744]: typo - it is any not model

**Comment by serengil** [comment]
> > Did you convert this weights to onnx? - https://github.com/serengil/deepface/blob/master/deepface/models/facial_recognition/Facenet.py#L1688
> > Should I wait same accuracy?
> 
> yes i have used this weights to convert it into onnx

Would you please run this unit test

https://github.com/serengil/deepface/blob/master/tests/test_verify.py#L16

with configuring variables as

```python
models = ["Facenet512ONNX"]
metrics = ["cosine", "euclidean", "euclidean_l2"]
detectors = ["opencv"]
```

wonder its score on unit test items.

File: tests/test_verify.py [hunk 5]
```diff
@@ -8,9 +8,9 @@
 
 logger = Logger()
 
-models = ["VGG-Face", "Facenet", "Facenet512", "ArcFace", "GhostFaceNet"]
+models = ["Facenet512ONNX"]
```
Comment (serengil) [inline tests/test_verify.py:11]: Please restore this file.

**Review by serengil — CHANGES_REQUESTED** [review]

**Comment by ShivamSinghal1** [comment]
> > > Did you convert this weights to onnx? - https://github.com/serengil/deepface/blob/master/deepface/models/facial_recognition/Facenet.py#L1688
> > > Should I wait same accuracy?
> > 
> > 
> > yes i have used this weights to convert it into onnx
> 
> Would you please run this unit test
> 
> https://github.com/serengil/deepface/blob/master/tests/test_verify.py#L16
> 
> with configuring variables as
> 
> ```python
> models = ["Facenet512ONNX"]
> metrics = ["cosine", "euclidean", "euclidean_l2"]
> detectors = ["opencv"]
> ```
> 
> wonder its score on unit test items.
> 
> ps: don't change unit test file in the git, just do it locally please

Accuracy is not matching with euclidean, do you know what could be the reason?

**Comment by serengil** [comment]
> Accuracy is not matching with euclidean, do you know what could be the reason?

Converting weights caused it most probably. What is the accuracy score you got?
